### ISSUE TITLE
tensorflow2 Custom training fine tuning issue

### ISSUE THREAD

**Issue by lovejing0306** [tensorflow/tensorflow#32583]
**I used  ResNet101V2 pretrained model training my dataset.**
Below is my code:

```
    batch_size = 16
    buffer_size = 5000
    image_size = 224
    classes = 9
    num_epochs = 200
    learning_rate = 0.000
    train_image_paths, train_image_labels = util.get_image_paths_and_labels(train_dir)
    train_dataset = tf.data.Dataset.from_tensor_slices((train_image_paths, train_image_labels))
    train_dataset = train_dataset.map(
        lambda path, label: weather.load_and_preprocess_image_and_label(path, label, image_size))
    train_dataset = train_dataset.shuffle(buffer_size=buffer_size)
    train_dataset = train_dataset.batch(batch_size=batch_size)
    val_image_paths, val_image_labels = util.get_image_paths_and_labels(val_dir)
    val_dataset = tf.data.Dataset.from_tensor_slices((val_image_paths, val_image_labels))
    # val_dataset = weather.load_data(val_tfrecords_path)
    val_dataset = val_dataset.map(
        lambda path, label: weather.load_and_preprocess_image_and_label(path, label, image_size))
    val_dataset = val_dataset.batch(batch_size=batch_size)
    base_model = ResNet101V2(input_shape=(image_size, image_size, 3),
                             include_top=False,
                             weights='imagenet')  # 'imagenet'
    base_model.trainable = True
    
    model = keras.Sequential([
        base_model,
        keras.layers.GlobalAveragePooling2D(),
        keras.layers.Dense(classes, activation='softmax', name='predictions')
    ])
    model.summary()
    
    loss_fn = keras.losses.CategoricalCrossentropy()
    # optimizer = keras.optimizers.Adam(learning_rate=learning_rate)
    optimizer = keras.optimizers.RMSprop(learning_rate=learning_rate)
    
    train_loss = keras.metrics.Mean()
    train_accuracy = keras.metrics.CategoricalAccuracy()
    
    val_loss = keras.metrics.Mean()
    val_accuracy = keras.metrics.CategoricalAccuracy()

    for epoch in range(num_epochs):
        for step, (images, labels) in enumerate(train_dataset):
            labels = tf.one_hot(labels, depth=classes)
            with tf.GradientTape() as tape:
                y_ = model(images)
                loss = loss_fn(labels, y_)
            grads = tape.gradient(loss, model.trainable_variables)

            optimizer.apply_gradients(zip(grads, model.trainable_variables))

            train_loss(loss)
            train_accuracy(labels, y_)  # model(images)
            if step % 50 == 0:
                print("Step {:03d}: TrainLoss: {:.3f}, TrainAccuracy: {:.3%}".format(step,
                                                                                     train_loss.result(),
                                                                                     train_accuracy.result()))
        # Display metrics at the end of each epoch
        print("Epoch {:03d}: TrainLoss: {:.3f}, TrainAccuracy: {:.3%}".format(epoch,
                                                                              train_loss.result(),
                                                                              train_accuracy.result()))
        train_loss.reset_states()
        train_accuracy.reset_states()

        for images, labels in val_dataset:
            labels = tf.one_hot(labels, depth=classes)
            y_ = model(images)
            loss = loss_fn(labels, y_)
            val_loss(loss)
            val_accuracy(labels, y_)
        print("Epoch {:03d}: ValLoss: {:.3f}, ValAccuracy: {:.3%}".format(epoch,
                                                                          val_loss.result(),
                                                                          val_accuracy.result()))
        val_loss.reset_states()
        val_accuracy.reset_states()
```

I set base_model.trainable = True, the training result is bad. 

```
    Step 000: TrainLoss: 2.219, TrainAccuracy: 18.750%
    Step 050: TrainLoss: 2.122, TrainAccuracy: 21.446%
    Step 100: TrainLoss: 2.134, TrainAccuracy: 20.854%
    Step 150: TrainLoss: 2.139, TrainAccuracy: 19.826%
    Step 200: TrainLoss: 2.140, TrainAccuracy: 19.652%
    Step 250: TrainLoss: 2.137, TrainAccuracy: 20.045%
    Step 300: TrainLoss: 2.133, TrainAccuracy: 20.328%
    Step 350: TrainLoss: 2.133, TrainAccuracy: 20.495%
    Epoch 005: TrainLoss: 2.134, TrainAccuracy: 20.396%
    Epoch 005: ValLoss: 2.137, ValAccuracy: 19.844%
    2019-09-17 10:38:11.732802: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:143] Filling 
    up shuffle buffer (this may take a while): 1707 of 5000
    2019-09-17 10:38:21.722296: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:143] Filling 
    up shuffle buffer (this may take a while): 3409 of 5000
    2019-09-17 10:38:31.856501: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:143] Filling 
    up shuffle buffer (this may take a while): 4311 of 5000
    2019-09-17 10:38:41.166725: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:193] 
    Shuffle buffer filled.
    Step 000: TrainLoss: 2.057, TrainAccuracy: 18.750%
    Step 050: TrainLoss: 2.102, TrainAccuracy: 21.324%
    Step 100: TrainLoss: 2.123, TrainAccuracy: 20.050%
    Step 150: TrainLoss: 2.124, TrainAccuracy: 20.530%
    Step 200: TrainLoss: 2.127, TrainAccuracy: 19.963%
    Step 250: TrainLoss: 2.130, TrainAccuracy: 20.568%
    Step 300: TrainLoss: 2.132, TrainAccuracy: 20.619%
    Step 350: TrainLoss: 2.132, TrainAccuracy: 20.388%
```

However，I use keras training way

```
        model.compile(loss='sparse_categorical_crossentropy',
                      optimizer=tf.keras.optimizers.RMSprop(lr=learning_rate),
                      metrics=['accuracy'])

        model.fit(train_dataset,
                  epochs=num_epochs,
                  validation_data=val_dataset)
```

the training result is good. 

I am very confused，anyone help me？
Thanks！

**Comment by gadagashwini** [comment]
> the training result is good.
> 
> I am very confused，anyone help me？
Please let us know what exactly the issue with training. Thanks!

**Comment by lovejing0306** [comment]
> > the training result is good.
> > I am very confused，anyone help me？
> 
> Please let us know what exactly the issue with training. Thanks!

I fine tuning the model in a custom training way and trained around 30 epochs, the train accuracy always around 20% . But I fine tuning the model in a tf.keras way (compile and fit), the train accuracy is right.

**Comment by gadagashwini** [comment]
@lovejing0306,
Did yo go through the custom training in [Tensorflow website](https://www.tensorflow.org/tutorials/eager/custom_training_walkthrough). Thanks!

**Comment by lovejing0306** [comment]
> @lovejing0306,
> Did yo go through the custom training in [Tensorflow website](https://www.tensorflow.org/tutorials/eager/custom_training_walkthrough). Thanks!

I copy the program from Tensorflow2 website, but it is not work.
Can you help me ? Thanks!

**Comment by gadagashwini** [comment]
@lovejing0306,
Can you please provide us the link from where you have copied the code. Thanks!

**Comment by lovejing0306** [comment]
> @lovejing0306,
> Can you please provide us the link from where you have copied the code. Thanks!

I have solved this problem. Thanks!
